cylc/uiserver/utils.py

```diff
@@ -14,6 +14,31 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 
+from typing import TYPE_CHECKING
+
+from jupyter_server.auth.identity import PasswordIdentityProvider
+
+if TYPE_CHECKING:
+    from cylc.uiserver.handlers import CylcAppHandler
+    from jupyter_server.auth.identity import (
+        IdentityProvider as JPSIdentityProvider,
+    )
+
+
+def is_token_authenticated(handler: 'CylcAppHandler') -> bool:
+    """Returns True if this request is bearer token authenticated.
+
+    E.G. The default single-user token-based authenticated.
+
+    In these cases the bearer of the token is awarded full privileges.
+    """
+    identity_provider: 'JPSIdentityProvider' = (
+        handler.serverapp.identity_provider  # type: ignore[union-attr]
+    )
+    return identity_provider.__class__ == PasswordIdentityProvider
```

Comment:
Why is it `True` if using `PasswordIdentityProvider`? I'm not familiar with the jupyter_server API but `PasswordIdentityProvider` suggests not using a token.

Also I have spotted `jupyter_server.auth.identity.IdentityProver` has its own `is_token_authenticated` method, is there a reason we are not using this? - https://github.com/jupyter-server/jupyter_server/blob/da948f050455a58d7afe07665ac2d777ee5bb2ff/jupyter_server/auth/identity.py#L489

Reply:
Passwords and tokens are both equivalent i.e. there is bearer token (or password) which grants full access to the server.

I'll look into the `is_token_authenticated` method. Because of the strange way Jupyter Hub wraps Jupyter Server I'm not sure whether it's safe to rely on it or not.

Reply:
@MetRonnie 

tldr; No, the jupyter_server `is_token_authenticated` method does not return the right results for this usage.

```
>>> PasswordIdentityProvider().is_token_authenticated(SimpleNamespace(current_user='me'))
False
```

This cylc-uiserver implementation is focused on determining whether the session is "bearer token" (more general than just tokens per-se) authenticated (passwords are just another form of bearer token) or whether authentication is being provided by an external service.

If authentication is bearer token based, then the bearer of the token should have full permissions to the server, that is to say, fine grained permissions do not apply as there is only one level of user. The safest way to implement this (that I could come up with) is to whitelist only the one authentication provider class that we want this rule to apply to.

Long term, it may be worth raising this with Jupyter Server as I think multi-user server access may be off their radar.

Reply:
Perhaps rename our method to `is_bearer_token_authenticated` or `is_bearer_authenticated` to avoid confusion?

Reply:
Changed and added more info into the docstring.